pkg/controller/core/traits/manualscalertrait/manualscalertrait_controller.go

```diff
@@ -143,42 +163,75 @@ func (r *Reconciler) scaleChildResources(ctx context.Context, mLog logr.Logger,
+		return util.ReconcileWaitResult,
+			util.PatchCondition(ctx, r, &manualScalar, cpv1alpha1.ReconcileError(errors.Wrap(err, errQueryOpenAPI)))
+	}
+	document, err := openapi.NewOpenAPIData(schema_doc)
+	if err != nil {
+		return util.ReconcileWaitResult,
+			util.PatchCondition(ctx, r, &manualScalar, cpv1alpha1.ReconcileError(errors.Wrap(err, errQueryOpenAPI)))
+	}
 	for _, res := range resources {
-		if res.GetKind() == util.KindDeployment && res.GetAPIVersion() == appsv1.SchemeGroupVersion.String() {
+		if locateReplicaField(document, res) {
 			found = true
 			resPatch := client.MergeFrom(res.DeepCopyObject())
-			mLog.Info("Get the deployment the trait is going to modify",
-				"deploy name", res.GetName(), "UID", res.GetUID())
+			mLog.Info("Get the resource the trait is going to modify",
+				"resource name", res.GetName(), "UID", res.GetUID())
 			cpmeta.AddOwnerReference(res, ownerRef)
 			unstructured.SetNestedField(res.Object, int64(manualScalar.Spec.ReplicaCount), "spec", "replicas")
-			// merge patch to scale the deployment
+			// merge patch to scale the resource
 			if err := r.Patch(ctx, res, resPatch, client.FieldOwner(manualScalar.GetUID())); err != nil {
-				mLog.Error(err, "Failed to scale a deployment")
+				mLog.Error(err, "Failed to scale a resource")
 				return util.ReconcileWaitResult,
-					util.PatchCondition(ctx, r, &manualScalar, cpv1alpha1.ReconcileError(errors.Wrap(err, errScaleDeployment)))
+					util.PatchCondition(ctx, r, &manualScalar, cpv1alpha1.ReconcileError(errors.Wrap(err, errScaleResource)))
 			}
-			mLog.Info("Successfully scaled a deployment", "UID", res.GetUID(), "target replica",
-				manualScalar.Spec.ReplicaCount)
+			mLog.Info("Successfully scaled a resource", "resource GVK", res.GroupVersionKind().String(),
+				"res UID", res.GetUID(), "target replica", manualScalar.Spec.ReplicaCount)
 		}
 	}
 	if !found {
-		mLog.Info("Cannot locate any deployment", "total resources", len(resources))
+		mLog.Info("Cannot locate any resource", "total resources", len(resources))
 		return util.ReconcileWaitResult,
-			util.PatchCondition(ctx, r, &manualScalar, cpv1alpha1.ReconcileError(fmt.Errorf(errLocateDeployment)))
+			util.PatchCondition(ctx, r, &manualScalar, cpv1alpha1.ReconcileError(fmt.Errorf(errScaleResource)))
 	}
 	return ctrl.Result{}, nil
 }
 
+// locateReplicaField call openapi RESTFUL end point to fetch the schema of a given resource and try to see
+// 	if it has a spec.replicas filed that is of type integer. We will apply duck typing to modify the fields there
+//  assuming that the fields is used to control the number of instances of this resource
+//  NOTE: This only works if the resource CRD has a structural schema, all `apiextensions.k8s.io/v1` CRDs do
+// https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#specifying-a-structural-schema
+func locateReplicaField(document openapi.Resources, res *unstructured.Unstructured) bool {
```

Comment:
I don't really like this way, what in my mind is we can maintain a map in traitDefinition like this:

```
apiVersion: core.oam.dev/v1alpha2
kind: TraitDefinition
metadata:
  name: manuelscalers.core.oam.dev
spec:
  definitionRef:
    name: manuelscalers.core.oam.dev
  appliesToWorkloads:
    - deployment.apps
  ducktypeMappings:
    - workloadName: deployment.apps
      mappings:
        - from: spec.replicaCount
          to: spec.replica
```

So we don't make the assumption(spec.replicas) here.

Reply:
I wonder who will add the mapping here?  An application developer or operator or the platform builder?

Reply:
The Platform builder who provision this trait. When a new Workload come in, the trait will add appliesToWorkloads support along with ducktypeMappings.

Reply:
The reason for the duck typing is to avoid too many manual steps. Granted, it won't apply to all cases but it can complement the manual mapping.